### ISSUE TITLE
Simulated runs on discrete search spaces

### ISSUE THREAD

**Issue by ArlindKadra** [awslabs/syne-tune#505]
**Is your feature request related to a problem? Please describe.**

I am observing the example provided in: https://github.com/awslabs/syne-tune/blob/790e3647a4e71d8f6ebfba608c26501a533d9c2a/examples/launch_simulated_benchmark.py#L55-L63 That simulates a run for a particular baseline on a given benchmark.  I was wondering if it would be possible to actually run on a limited search space that has a fixed set of hyperparameter candidates without using a surrogate, probably allowing for a grid initialization of the search space:

https://github.com/awslabs/syne-tune/blob/790e3647a4e71d8f6ebfba608c26501a533d9c2a/examples/launch_simulated_benchmark.py#L38-L41

**Describe the solution you'd like**

Limiting the number of considered hyperparameter candidates to make use of the already existing benchmarks inside the repository without fantasizing the performance for out of sample hyperparameter configurations.

**Describe alternatives you've considered**

NA

**Additional context**

NA

**Comment by mseeger** [comment]
Hello. Yes, this is something we can support, I'll look into this.

We have code for such a restricted search in place, also for the methods using GP-based BO, it is just we so far did not create explicit links between the blackbox and the scheduler, which is what we have to do here.

**Comment by mseeger** [comment]
https://github.com/awslabs/syne-tune/pull/524

This PR should solve this issue

**Comment by mseeger** [comment]
This has been merged, and docs have been updated.

**Comment by ArlindKadra** [comment]
I tried the code with ASHA given a custom benchmark of ours and it worked fine and the method was competitive. However, when I switched methods to DEHB or BOHB the script crashed, I think the patch might not be integrated for the methods.

Would it be possible to integrate the patch for the other methods too? 

BOHB should be easy I think, while, DEHB is slightly more complex with the mutations. The shortest way would be to match after the mutation with the closest evaluated hp configuration in the vicinity.

**Comment by mseeger** [comment]
Hello @ArlindKadra . You are correct, this does not work with BOHB or DEHB at the moment. For BOHB, this is on our list to do.

**Comment by mseeger** [comment]
For DEHB, I am hesitant doing this, because you can already right now do what you suggested, by defining 1-NN as surrogate.

If you look at `benchmarking/commons/benchmark_definitions/lcbench.py`, you see how to choose this (`surrogate` and `surrogate_kwargs`).

**Comment by mseeger** [comment]
Note that you can always define such a surrogate. That would deal with the problem in the benchmark itself.

What I implemented based on your request, was a means to deal with the finite-ness of the surrogate in the scheduler itself. But if this needs things like nearest neighbor matching or interpolation, this is already done well (and fast) in the blackbox repository.

**Comment by ArlindKadra** [comment]
Thanks for the answer @mseeger . I will take a look at the surrogate as soon as possible. I might be wrong on my understanding, but I am going to leave a remark before verifying it.

I think my understanding and notion differs a bit, if you map to the closest hyperparameter configuration, while, keeping the hyperparameter configurations restricted, you are still running on the benchmark. If I use a surrogate, to fantasize the performance of a hyperparameter configuration in a neighborhood of the evaluated hyperparameter configuration, then, I am not running on the benchmark anymore.

**Comment by mseeger** [comment]
Well, you can think of the benchmark together with the surrogate as a benchmark itself, extended over the full space. You are right, in that different surrogates would then result in different benchmarks.

But take the case of DEHB. I cannot force DEHB to only consider configs from a finite set, because it works with encoded vectors. But when I use 1-NN surrogate, I am doing what you suggested. Say, DEHB suggests a vector x. First, I map this back to a config c(x) by rounding. Next, the surrogate maps this to the NN c'(x), for which I have a target value.

If you use 1-NN, you do not come up with new values, in the sense of interpolation.

**Comment by ArlindKadra** [comment]
Ah, I see, this is nearly what I had in mind.

What I was thinking was the following: you have the array of configurations, with all their associated array of performances. You push all your configurations once at the beginning to the unit cube space for DEHB, when it suggests a configuration you map it to the closest configuration in the unit cube space from the transformed list of original candidates (1-NN), and from there you already have the label from the associated array of performances by just checking the index.  This would remove the intermediate mapping to c(x).

**Comment by mseeger** [comment]
Do you think this would be a large difference? If possible, I'd like to keep this simple.

**Comment by ArlindKadra** [comment]
No, I agree with you, I think the rounding will still map to a close valid config, and with the 1-NN, it will still match to what is available from the restricted config space.

**Comment by ArlindKadra** [comment]
On another note, I am having a bit of trouble having reproducible results.  I am passing a seed number here:

https://github.com/awslabs/syne-tune/blob/4ca3afd2cb219aba6dc7f1c9955e1a978566b562/syne_tune/blackbox_repository/simulated_tabular_backend.py#L418

Although, unnecessary as I only have 1 seed.

and here:
https://github.com/awslabs/syne-tune/blob/4ca3afd2cb219aba6dc7f1c9955e1a978566b562/syne_tune/optimizer/baselines.py#L163

Is there something I might be doing wrong? I also tried DEHB and I could not get reproducible results.
I followed the seed passing for ASHA and it looked ok, it is being passed through and a random state is created, not sure if the random state is used to sample the initial configurations though.

**Comment by mseeger** [comment]
Hello,
the `seed` in the blackbox repo is if the blackbox includes different samples. If that is the case, you need to indeed fix this seed to get reproducible results.

But there is also `random_seed` passed to the scheduler. This is the "master random seed" which controls all the sampling. If you do not set that one, it is drawn at random and printed, but this will give you different results then.

**Comment by mseeger** [comment]
If you use the `benchmarking` formalism, use `--random_seed <yourseed>`.

**Comment by mseeger** [comment]
That should give the same results for schedulers that rely on random sampling (like DEHB). But for BO-based schedulers, you'll get some variance due to the optimization over hyperparameters. But it should be small.

**Comment by ArlindKadra** [comment]
> Hello, the `seed` in the blackbox repo is if the blackbox includes different samples. If that is the case, you need to indeed fix this seed to get reproducible results.

Thanks for the answer. Yes, I also understood the same and I am giving that too, although my objective evaluations only have 1 seed. More for good practice.

> But there is also `random_seed` passed to the scheduler. This is the "master random seed" which controls all the sampling. If you do not set that one, it is drawn at random and printed, but this will give you different results then.

I am also passing the `random_seed`, however, I cannot replicate my results, neither with DEHB, nor with ASHA. Are you able to replicate results in your environment?

**Comment by mseeger** [comment]
Yes, usually this works for me.

Could you open another issue and provide the code you are using? I can also look at a branch.

**Comment by mseeger** [comment]
Hello @ArlindKadra , I just checked the code for BOHB, which has been implemented by Aaron.
In fact, that code resembles DEHB, in that when proposing a new config, it samples encoded vectors in some particular way (I suppose, to implicitly optimize the EI acq func), and then converts the best vector back to a config.

Again, I am hesistant changing this. My feeling is this is best done in the same way as BOHB, namely with a 1-NN surrogate for the blackbox.

Does this work for you?
